lib/symboldatabase.cpp

```diff
@@ -1643,61 +1685,57 @@ void SymbolDatabase::createSymbolDatabaseExprIds()
-                }
-            } else if (isCPP() && Token::simpleMatch(tok, "this")) {
-                if (thisId == 0)
-                    thisId = id++;
-                tok->exprId(thisId);
-            }
-        }
-
-        // Apply CSE
-        for (const auto& p:exprs) {
-            const std::vector<Token*>& tokens = p.second;
-            const std::size_t N = tokens.size();
-            for (std::size_t i = 0; i < N; ++i) {
-                Token* const tok1 = tokens[i];
-                for (std::size_t j = i + 1; j < N; ++j) {
-                    Token* const tok2 = tokens[j];
-                    if (tok1->exprId() == tok2->exprId())
-                        continue;
-                    if (!isSameExpression(isCPP(), true, tok1, tok2, mSettings.library, false, false))
-                        continue;
-                    nonneg int const cid = std::min(tok1->exprId(), tok2->exprId());
-                    tok1->exprId(cid);
-                    tok2->exprId(cid);
+                if (Token::Match(tok, "%name% <") && tok->next()->link()) {
+                    baseIds[tok->str()] = id;
+                    tok->exprId(id);
+                    ++id;
+                } else {
+                    const auto it = baseIds.find(tok->str());
+                    if (it != baseIds.end()) {
+                        tok->exprId(it->second);
+                    } else {
+                        baseIds[tok->str()] = id;
+                        tok->exprId(id);
+                        ++id;
+                    }
                 }
+
+                setParentExprId(tok, exprIdMap, id);
             }
         }
+
         // Mark expressions that are unique
-        std::unordered_map<nonneg int, Token*> exprMap;
+        std::vector<std::pair<Token*, int>> uniqueExprId(id);
         for (Token* tok = const_cast<Token*>(scope->bodyStart); tok != scope->bodyEnd; tok = tok->next()) {
-            if (tok->exprId() == 0)
+            const auto id2 = tok->exprId();
+            if (id2 == 0)
                 continue;
-            auto p = exprMap.emplace(tok->exprId(), tok);
-            // Already exists so set it to null
-            if (!p.second) {
-                p.first->second = nullptr;
-            }
+            uniqueExprId[id2].first = tok;
+            uniqueExprId[id2].second++;
         }
-        for (const auto& p : exprMap) {
-            if (!p.second)
+        for (const auto& p : uniqueExprId) {
+            if (!p.first || p.second != 1)
                 continue;
-            if (p.second->variable()) {
-                const Variable* var = p.second->variable();
-                if (var->nameToken() != p.second)
+            if (p.first->variable()) {
+                const Variable* var = p.first->variable();
+                if (var->nameToken() != p.first)
                     continue;
             }
-            p.second->setUniqueExprId();
+            p.first->setUniqueExprId();
```

Comment:
@pfultz2 Could you explain a little how exprid is supposed to work. Why do we have a special "unique" flag?
If there is a expression like: `x = 0;` then there will be problems if the `0` gets a expression id if that is by intention then please explain the philosophy of that..

Reply:
it seems that it's best to not set the exprid for any leafs is that correct?

I can see that the old implementation even set exprid for some "(" tokens that do not have astChildren or astParent. Is that by design?

Reply:
> Why do we have a special "unique" flag?

The unique flag means there is no other expression that is the same. We use that in ValueFlow to skip forwarding values since we wont ever find a matching expression.

> If there is a expression like: x = 0; then there will be problems if the 0 gets a expression id 

I dont think literals were meant to have expression ids. I think thats waht `isExpression` filters out.

> it seems that it's best to not set the exprid for any leafs is that correct?

We shouldn't set it for literals, but we should set it for variables(which I consider leafs) since references can have the same expression id but different varid.

> I can see that the old implementation even set exprid for some "(" tokens that do not have astChildren or astParent. Is that by design?

Probably not.